docs/src/tutorial.rst

```diff
@@ -0,0 +1,487 @@
+
+.. code-block:: ts
+
+    new widget_service.WidgetService(this, 'Widgets');
+
+Make sure it builds and creates a stack
+(we don't show the stack as it's almost 300 lines).
+
+.. code-block:: sh
+
+    npm run build
+    cdk synth
+
+.. _deploy_and_test:
+
+Step 5: Deploy and Test the App
+===============================
+
+Before you can deploy your first |cdk| app,
+you must bootstrap your deployment,
+which creates some AWS infracture that the |cdk|
+needs.
+See the **bootstrap** section of the :doc:`tools` topic for details.
+
+.. code-block:: sh
+
+    cdk bootstrap
+
+Run the following command to deploy your app.
+
+.. code-block:: sh
+
+    cdk deploy
+
+If the deployment is successfull,
+save the URL for your server, which appears in the last line in the window,
+where GUID is an alpha-numeric GUID and REGION is your region.
+
+.. code-block:: sh
+
+    https://GUID.execute-REGION.amazonaws.com/prod/
+
+You can test your app by getting the list of widgets (currently empty) by navigating to this URL in a
+browser or use the following **curl** command.
+
+.. code-block:: sh
+
+    curl -X GET 'https://GUID.execute-REGION.amazonaws.com/prod'
+
+You can also open the AWS Console,
+navigate to the API Gateway service,
+find **Widget Service** in the list.
+Select **GET** and **Test** to test the function.
+Since we haven't stored any widgets yet, the output should be similar to the following
+(there may be some slight differences in whitespace and quotation marks).
+
+.. code-block:: sh
+
+    { "widgets": [] }
+
+.. _adding_functions:
+
+Step 6: Add the Individual Widget Functions
+===========================================
+
+The next step is to create |LAM| functions to create, show, and delete
+individual widgets.
+Replace the existing **exports.main** function in *widgets.js* with the following code.
+
+.. code-block:: js
+
+    exports.main = async function(event, context) {
+      try {
+        var method = event.httpMethod;
+
+        if (method == "GET") {
+          if (event.path.length == 1) {
+            const data = await S3.listObjectsV2({ Bucket: bucketName }).promise();
+            var body = {
+              widgets: data.Contents.map(function(e) { return e.Key })
+            };
+            return {
+              statusCode: 200,
+              headers: {},
+              body: JSON.stringify(body)
+            };
+          }
+          else {
+            // GET /name
+            var name = event.path.substring(1, event.path.length);
+  
+            const data = await S3.getObject({ Bucket: bucketName, Key: name}).promise();
+            var body = data.Body.toString('utf-8');
+
+            return {
+              statusCode: 200,
+              headers: {},
+              body: JSON.stringify(body)
+            };
+          }
+        }
+        else if (method == "POST") {
+          // POST /name
+          // Make sure we have a name
+          if (event.path.length == 1) {
```

Comment:
===

But actually, why look at length? Why not just say:

```ts
if (event.path === '/') {
}
```

Because that's literally what you're trying to detect, except you're going about it in a roundabout way. What you're doing here is "if my string is 1 character long, then it MUST be a single slash!". To me, that seems to rest on a number of assumptions that are shaky until established otherwise. Also, there's no need to make those assumptions, because you can just write what you intend to say.

In fact, I would probably write the following:

```ts
// At the top: strip the leading slash if present
var widgetName = event.path.startsWith('/') ? event.path.substring(1) : event.path;

// Then, at every location where you expect a widget name
if (!widgetName) { 
  return { ... error response ... };
}
// continue to use widgetName...
```